[feature] Snapping also snaps to the currently digitised feature

olivierdalang opened the pull request:
Snapping now also snaps to the currently digitised feature (a.k.a. rubberband).

This opens some really nice cad-like editing abilities, such as closing features at 90° angle or precisely align segments inside a feature (for U or T shapes).  It's very efficient once one is used to the alt+a shortcut to lock the angle.

![pr_self-snapping](https://user-images.githubusercontent.com/1894106/81065380-a37bc100-8edb-11ea-8aed-6fd52d843a26.gif)

I implemented it in QgsMapCanvasSnappingUtils, where the new logic is triggered if the active map tool is a QgsMapToolCapture, so that it should work with all the advanced digitizing tools as well.

As with the other PR, I open this as a draft for feedback before going too far in a possibly wrong direction. I haven't looked into tests yet, will see how it's done for regular snapping.

UX question : currently, this is always enabled. Are there scenarios where self-snapping would be an annoyance ? It would be nice if we could avoid one more snap setting.

Implementation question : I'm not sure this is the best approach, as I had to re-implement some logic (e.g. snapping on segments, and priority of vertices over edges). But the existing classes (QgsPointLocator) seems to only work with a layer instance, which we don't have in this case. So any suggestion is welcome !

Cheers

olivierdalang committed:
[feature][selfsnap] add snapping to currently digitized feature for capture tool

Saijin-Naib commented:
This is AMAZING

nyalldawson commented:
> Are there scenarios where self-snapping would be an annoyance ? It would be nice if we could avoid one more snap setting.

Yes, unfortunately I think there is going to be cases when self-snapping will get in the way...

olivierdalang commented:
Ok I added a new snap setting, so the feature can be toggled on/off as needed.

@wonder-sk As you've put a lot of work in snapping, do you have an opinion about the implementation ? Would you see a clean way to integrate this into QgsSnappingUtils, to take advantage of all the existing logic (incl. for instance snapping on intersections or indexing if working on a very large feature) ?

m-kuhn commented on src/core/qgssnappingconfig.h:
```diff
@@ -335,6 +335,12 @@ class CORE_EXPORT QgsSnappingConfig
     //! Sets if the snapping on intersection is enabled
     void setIntersectionSnapping( bool enabled );
 
+    //! Returns if self snapping is enabled
```
Can you add a minimal description of what "self snapping" is and add a `\since QGIS 3.14`

wonder-sk commented:
Really nice stuff - this has been missing for a long time :-)

Implementation-wise, I would do it in a slightly different way: I would modify QgsSnappingUtils to temporarily store an extra geometry (or multiple geometries?) - so there would be two new functions e.g.:
- void setTemporaryExtraSnapGeometry( const QgsGeometry &geom, const QgsCoordinateReferenceSystem &crs )
- void resetTemporaryExtraSnapGeometry()

Then capture tool would update the extra geometry when the capture curve would be updated and the snapping utils class would implement all the logic to allow snap on such geometry. The advantage would be that this would not be limited only to the QgsMapToolCapture-based map tools, but any map tool could make use of that as needed... Also, given it would be implemented in qgis_core instead of qgis_gui this would be even more flexible.

nirvn commented:
Nice feature. The code looks good as is yet it'd be a shame to merge this now to meet feature freeze deadline and not take the extra time to implement Martin's sound suggestions.

Could we agree to a feature freeze exception right away to give time for the author to do the relevant changes?

olivierdalang commented:
@wonder-sk Thanks for the input !! The thing I'm not sure though is how to deal with this approach in terms of UI, if the extra-snap is of such general purpose, as it's not really known in advance for what purpose the extra snap geometry will be used.

The self-snapping toggle would actually becomes more of a QgsMapToolCapture property than a QgsSnapping property... so that the button wouldn't really fit in the snapping toolbar anymore (how would we label the button ? and what icon ?)

Or should we restrict the setTemporaryExtraSnapGeometry to something like setActiveToolSnapGeometry, so that it's clear the intended use is for map tools ? Do you have other scenarios in mind in which case we may want extra snapping where it's not about a map tool ? I can think about snapping to custom grid, but then you'll want to enable/disable that separately as well.

If it's OK to restrict to the map tool use case, there would be another approach, closer to what I made, that would allow similar flexibility, and that may be more clear in terms of API. We could just create a SnappableToolinterface, where SnappableTooltools implement a method returning a geometry. Instead of filtering on QgsMapToolCapture in lines https://github.com/qgis/QGIS/pull/36200/files#diff-49c481af803042cad76ecf484e6b40f4R104-R106 we would filter on SnappableTool.

IMO it may be better to start like this, then add more flexibility later on, when we actually have an use case. What do you think ?

(Also, other issue is FF is today, and I'll be out of office in 30 minutes :-) I'm available and willing to polish this out before release though !)

wonder-sk commented:
> The self-snapping toggle would actually becomes more of a QgsMapToolCapture property than a QgsSnapping property... so that the button wouldn't really fit in the snapping toolbar anymore (how would we label the button ? and what icon ?)

This may not be necessarily a problem... We can think of "self-snapping enabled" as a preference for map tools that support it. There is already "avoid intersections" flag in snapping configuration which has even nothing to do with snapping :-) and it is only supported by some tools. So I can imagine that self-snapping could be similar, if we wanted to improve UX, we could enable/disable the button based on whether this is supported by the active map tool or something like that.

> Or should we restrict the setTemporaryExtraSnapGeometry to something like setActiveToolSnapGeometry, so that it's clear the intended use is for map tools ? Do you have other scenarios in mind in which case we may want extra snapping where it's not about a map tool ?

Yeah I was half-minded how to call this concept :-) One other scenario I had in mind is for example when implementing some algorithm that uses snapping and it would be handy to have some extra snapping points there for some reason (but I have no concrete example in mind). Your suggestion for the name (setActiveToolSnapGeometry) also sounds fine to me.

> If it's OK to restrict to the map tool use case, there would be another approach, closer to what I made, that would allow similar flexibility, and that may be more clear in terms of API. We could just create a SnappableToolinterface, where SnappableTooltools implement a method returning a geometry.

Yeah, that's yet another approach that could work. Just maybe instead of having it tied to map tools, it would be nice if it could be more generic - e.g. SnapGeometryProvider interface with a method to return geometry, and it would be added/removed when a particular map tool gets activated/deactivated. I don't have a strong opinion which route to take - as long as the logic is in qgis_core and not just in qgis_gui.

> I'm available and willing to polish this out before release though ! I could probably finish this next week, depending on what implementation we agree on

I am personally fine with that to merge it next week - it is relatively small piece in terms of the impact on the code, yet adds a lot of value!

olivierdalang committed:
fix #36585 where snapping on segments with a constraint snapped to edges instead of vertices

olivierdalang committed:
show snap indicators for snaps with constraints

olivierdalang commented:
Thanks for the FF exemption :-)

@wonder-sk
So I gave it another shot following your suggestion, but instead of `setTemporaryExtraSnapGeometry` and `resetTemporaryExtraSnapGeometry`, I implemented `addExtraSnappingLayer` and `removeExtraSnappingLayer`. It would typically be used with memory layers.

The reason for using layers instead of geometries is that is that this better integrates with the rest of the code, as `QgsPointLocator` expects layers. This allows to reuse all the code (caching, CRS conversion, etc.), and avoids needing to reinvent the wheel in how to store and maintain a collection of geometries with CRS. This means that `QgsMapToolCapture` now keeps a memory layer with it's rubberband.

What do you think ?

By the way, I also added a commit to fix #36585 (which is needed to close features properly, the main point of this PR) and another to show the snap point with a constraint (which currently snapped "silently"). Let me know if these should be in a separate PR ?

wonder-sk commented:
hi @olivierdalang 

This looks good to me! Few comments:
- it may be good to have a getter for the extra snap layers (for introspection)
- map tool capture could just update geometry (that would be always present) rather than doing delete+add (but that's maybe a micro-optimization)
- can we have some unit tests for this :-)

olivierdalang commented:
I added the getter, did the micro-optimisation and added some tests.

Not sure about the getter, is it OK like this, or should I use something a bit more complex to ensure the QSet isn't modified from outside ? (I saw `_qgis_listQPointerToRaw()` from `qgsmaplayerlistutils.h` is used elsewhere in the code, but also read some stuff about Qt's implicit sharing, which I don't fully get, but seems to indicate that this is already implicitly taken care of).

Many thanks again for the review @wonder-sk  :-)

Also please let me squash the commits before merging (so we don't pollute the git history too much)

uclaros commented:
This is sooooo nice!
A couple of issues I noticed though:
1. When digitizing a polygon, its last segment (between first point and last point) cannot be used for snapping.
2. When *Snap to Centroid* is enabled, on the first click of digitizing you get a `CRITICAL    Qt : GEOS exception: IllegalArgumentException: point array must contain 0 or >1 elements`
3. Topological points are not added to self when using a *Segment* or *Middle of segment* self snap.

olivierdalang committed:
[selfsnap] followup : snap to closing line of poly and fix geos error message

olivierdalang commented:
> 1. When digitizing a polygon, its last segment (between first point and last point) cannot be used for snapping.
> 2. When _Snap to Centroid_ is enabled, on the first click of digitizing you get a `CRITICAL    Qt : GEOS exception: IllegalArgumentException: point array must contain 0 or >1 elements`
> 3. Topological points are not added to self when using a _Segment_ or _Middle of segment_ self snap.

@uclaros 
1 & 2 are easy fixes, thanks for testing :-)

About 3, I'm not sure how easy that is to implement (not yet familiar with topological editing code). It may be difficult as the feature doesn't yet exist in the layer... If OK, I'd prefer to leave this as a future enhancement, as we're already over feature freeze, it already took longer than I expected on this feature, and... it's a bit of an edge case :-)

m-kuhn commented on src/app/qgssnappingwidget.cpp:
```diff
@@ -287,6 +287,14 @@ QgsSnappingWidget::QgsSnappingWidget( QgsProject *project, QgsMapCanvas *canvas,
+  mSelfSnappingAction = new QAction( tr( "Self-snapping" ), this );
+  mSelfSnappingAction->setCheckable( true );
+  mSelfSnappingAction->setIcon( QIcon( QgsApplication::getThemeIcon( "/mIconSnappingSelf.svg" ) ) );
+  mSelfSnappingAction->setToolTip( tr( "Enable Self-snapping" ) );
```
```suggestion
  mSelfSnappingAction->setToolTip( tr( "If self snapping is enabled, snapping will also use the current state of the digitized feature into consideration." ) );
```

I think the tool tip can be a bit more verbose.

m-kuhn commented on src/core/qgscadutils.h:
```diff
@@ -92,7 +92,13 @@ class CORE_EXPORT QgsCadUtils
       QgsPointXY finalMapPoint;
 
-      //! Snapped segment - only valid if actually used for something
+      //! Snapped point - only valid if actually used for something
```
```suggestion
      /**
       * Snapped point - only valid if actually used for something
       * \since QGIS 3.14
       */
```

m-kuhn commented on src/core/qgssnappingutils.h:
```diff
@@ -185,6 +185,48 @@ class CORE_EXPORT QgsSnappingUtils : public QObject
 
+    /**
+     * Supply an extra snapping layer (typically a memory layer).
+     * This is can be used by map tools to provide additionnal
```
```suggestion
     * This can be used by map tools to provide additionnal
```

m-kuhn commented on src/gui/qgsmaptoolcapture.cpp:
```diff
@@ -63,6 +72,8 @@ QgsMapToolCapture::~QgsMapToolCapture()
     mValidator->deleteLater();
     mValidator = nullptr;
   }
+  mExtraSnapLayer->deleteLater();
```
```suggestion
  mCanvas->snappingUtils()->removeExtraSnapLayer( mExtraSnapLayer );
  mExtraSnapLayer->deleteLater();
```

Or is this guaranteed to be taken care of elsewhere?

m-kuhn commented on src/app/qgssnappingwidget.cpp:
```diff
@@ -291,7 +291,7 @@ QgsSnappingWidget::QgsSnappingWidget( QgsProject *project, QgsMapCanvas *canvas,
   mSelfSnappingAction->setCheckable( true );
   mSelfSnappingAction->setIcon( QIcon( QgsApplication::getThemeIcon( "/mIconSnappingSelf.svg" ) ) );
-  mSelfSnappingAction->setToolTip( tr( "Enable Self-snapping" ) );
+  mSelfSnappingAction->setToolTip( tr( "If self snapping is enabled, snapping will also use the current state of the digitized feature into consideration." ) );
```
```suggestion
  mSelfSnappingAction->setToolTip( tr( "If self snapping is enabled, snapping will also take the current state of the digitized feature into consideration." ) );
```

m-kuhn commented on python/core/auto_generated/qgssnappingutils.sip.in:
```diff
@@ -166,6 +166,42 @@ Set if invisible features must be snapped or not.
+    void addExtraSnapLayer( QgsVectorLayer *vl );
+%Docstring
+Supply an extra snapping layer (typically a memory layer).
+This is can be used by map tools to provide additionnal
```
```suggestion
This can be used by map tools to provide additionnal
```

olivierdalang committed:
[selfsnap] apply pr review

m-kuhn commented:
Thanks @olivierdalang and @wonder-sk
